niceman/distributions/venv.py

```diff
@@ -0,0 +1,211 @@
+        venvs = []
+        for venv_path in venv_paths:
+            package_details, file_to_pkg = self._get_package_details(venv_path)
+            pkg_to_found_files = defaultdict(list)
+            for path in set(unknown_files):  # Clone the set
+                if path in file_to_pkg:
+                    unknown_files.remove(path)
+                    pkg_to_found_files[file_to_pkg[path]].append(
+                        os.path.relpath(path, venv_path))
+
+            packages = [VenvPackage(name=details["name"],
+                                    version=details["version"],
+                                    origin_location=details["location"],
+                                    files=pkg_to_found_files[name])
+                        for name, details in package_details.items()]
+
+            found_package_count += len(packages)
+
+            venvs.append(
+                VenvEnvironment(name=short_names[venv_path],
+                                path=venv_path,
+                                python_version=self._python_version(venv_path),
+                                packages=packages))
+
+        lgr.info("%s: %d packages with %d files, and %d other files",
+                 self.__class__.__name__,
+                 found_package_count,
+                 total_file_count - len(unknown_files),
+                 len(unknown_files))
+
+        if venvs:
+            yield (VenvDistribution(name="venv",
+                                    venv_version=self._venv_version(),
+                                    path=self._venv_exe_path(),
+                                    environments=venvs),
+                   list(unknown_files))
+
+    def _pip_packages(self, venv_path):
+        # We could use either 'pip list' or 'pip freeze' to get a list
+        # of packages.  The choice to use 'list' rather than 'freeze'
+        # is based on how they show editable packages.  'list' outputs
+        # a source directory of the package, whereas 'freeze' outputs
+        # a URL like "-e git+https://github.com/[...]".
+        #
+        # It would be nice to use 'pip list --format=json' rather than
+        # the legacy format.  However, currently (pip 9.0.1, 2018/01),
+        # the json format does not include location information for
+        # editable packages (though it is supported in a developmental
+        # version).
+        try:
+            out, _ = self._session.execute_command([venv_path + "/bin/pip",
+                                                    "list", "--format=legacy"])
+
+        except Exception as exc:
+            lgr.warning("Could determine pip packages for %s: %s",
+                        venv_path, exc_str(exc))
+            return
+        for pkg, _, loc in parse_pip_list(out):
+            yield pkg, loc
+
+    def _python_version(self, venv_path):
+        try:
+            out, err = self._session.execute_command(
+                venv_path + "/bin/python --version")
```

Comment:
here and in other spots, preferable to invoke via a list of `[cmd, *options]` not a string, to be able to test in obscure test environments with spaces in the path directory names

Reply:
good point